FIPS/fip-draft_foundational_fvm.md

```diff
@@ -0,0 +1,1010 @@
+
+### Context and VM interface
+
+In the Filecoin blockchain, the virtual machine (VM) is the component in the
+execution layer responsible for applying transactions over a state tree. These
+transactions are encapsulated in messages. There are two kinds of messages:
+
+- Explicit messages are sent by secp256k1 or BLS account actors, appear on-chain
+  when executed, and carry a signature to prove their authenticity. 
+- Implicit messages are generated by the system, do not appear on chain, and
+  they carry no signature. Examples include cron ticks and reward cycles.
+
+A VM is instantiated with these input parameters:
+
+```rust
+let machine = Machine::new(
+    epoch,             // chain epoch at which to execute messages
+    base_fee,          // base fee currently in effect
+    base_circ_supply,  // base circulating supply
+    nv,                // network version
+    state_root,        // root CID of the initial state tree
+    blockstore,        // blockstore to load and save state objects during execution
+)
+```
+
+Messages are then applied to the `machine`. Every application generates a return
+object containing at least a message receipt, a miner penalty, and a miner tip.
+Implementations may return extra data such as diagnostics info, backtraces, or
+gas traces.
+
+```rust
+let ApplyRet { receipt, penalty, tip } := machine.apply_message(msg1);
+let ApplyRet { receipt, penalty, tip } := machine.apply_message(msg2);
+let ApplyRet { receipt, penalty, tip } := machine.apply_message(msg3);
+```
+
+The message receipt is recorded on chain. It holds the exit code of the
+transaction, the gas used, and (optionally) return data only if successful.
+
+The penalty and the tip are values used by the cryptoeconomic rules of the
+protocol to calculate the block reward for the block producer.
+
+Once the caller has finished applying all messages, it flushes the machine. This
+computes the final state root. Depending on implementation, this operation may
+also commit newly written blocks to the blockstore.
+
+```rust
+let final_state_root = machine.flush();
+```
+
+With the introduction of the FVM, this VM API is largerly preserved, although
+implementations may opt to deviate from it.
+
+### Technology
+
+The proposed FVM execution runtime is [WebAssembly (WASM)].
+
+The chosen WASM architecture is variant is 32-bits, and not 64-bit. Reasons include:
+
+- Widespread support
+- Lower memory footprint
+- 4GiB addressable memory not deemed limiting for blockchain use cases
+- No loss of function: 64-bit integer types and operations are supported
+
+The FVM requires [multi-value support]. It explicitly forbids [SIMD
+instructions], floating number values and operations, and [threads] to prevent
+[Nondeterministic] behavior.
+
+### Responsibilities
+
+The FVM is responsible for these tasks:
+
+- Loading the code of invoked actors, turning them into WASM modules, and
+  optimizing the process through compilation and caching.
```

Comment:
> turning them into WASM modules

This is confusing, the code is _already_ a wasm module before it gets to the FVM. I believe you mean "loading on-chain wasm modules and compiling them down to native code"?

Reply:
Yep, I was thinking too close to the code, i.e. the instantiation of a `wasmtime::Module` from bytecode.